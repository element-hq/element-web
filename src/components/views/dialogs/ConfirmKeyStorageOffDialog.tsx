/*
Copyright 2025 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React from "react";
import ErrorIcon from "@vector-im/compound-design-tokens/assets/web/icons/error";
import { Button } from "@vector-im/compound-web";
import { PopOutIcon } from "@vector-im/compound-design-tokens/assets/web/icons";

import { _t } from "../../../languageHandler";
import defaultDispatcher from "../../../dispatcher/dispatcher";
import { EncryptionCard } from "../settings/encryption/EncryptionCard";
import { EncryptionCardButtons } from "../settings/encryption/EncryptionCardButtons";
import { type OpenToTabPayload } from "../../../dispatcher/payloads/OpenToTabPayload";
import { Action } from "../../../dispatcher/actions";
import { UserTab } from "./UserTab";

interface Props {
    onFinished: (dismissed: boolean) => void;
}

/**
 * Ask the user whether they really want to dismiss the toast about key storage.
 *
 * Launched from the {@link SetupEncryptionToast} in mode `TURN_ON_KEY_STORAGE`,
 * when the user clicks "Dismiss". The caller handles any action via the
 * `onFinished` prop which takes a boolean that is true if the user clicked
 * "Yes, dismiss".
 */
export default class ConfirmKeyStorageOffDialog extends React.Component<Props> {
    public constructor(props: Props) {
        super(props);
    }

    private onGoToSettingsClick = (): void => {
        // Open Settings at the Encryption tab
        const payload: OpenToTabPayload = {
            action: Action.ViewUserSettings,
            initialTabId: UserTab.Encryption,
        };
        defaultDispatcher.dispatch(payload);
        this.props.onFinished(false);
    };

    private onDismissClick = (): void => {
        this.props.onFinished(true);
    };

    public render(): React.ReactNode {
        return (
            <EncryptionCard
                Icon={ErrorIcon}
                destructive={true}
                title={_t("settings|encryption|confirm_key_storage_off")}
            >
                {_t("settings|encryption|confirm_key_storage_off_description", undefined, {
                    a: (sub) => (
                        <>
                            <br />
                            <a href="https://element.io/help#encryption5" target="_blank" rel="noreferrer noopener">
                                {sub} <PopOutIcon />
                            </a>
                        </>
                    ),
                })}
                <EncryptionCardButtons>
                    <Button onClick={this.onGoToSettingsClick} autoFocus kind="primary" className="">
                        {_t("common|go_to_settings")}
                    </Button>
                    <Button onClick={this.onDismissClick} kind="secondary">
                        {_t("action|yes_dismiss")}
                    </Button>
                </EncryptionCardButtons>
            </EncryptionCard>
        );
    }
}
