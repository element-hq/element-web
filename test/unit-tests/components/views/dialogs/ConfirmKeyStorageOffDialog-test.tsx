/*
Copyright 2025 New Vector Ltd.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import React from "react";
import { render } from "jest-matrix-react";

import ConfirmKeyStorageOffDialog from "../../../../../src/components/views/dialogs/ConfirmKeyStorageOffDialog";

describe("ConfirmKeyStorageOffDialog", () => {
    beforeEach(() => {
        jest.resetAllMocks();
    });

    it("renders", () => {
        const dialog = render(<ConfirmKeyStorageOffDialog onFinished={jest.fn()} />);
        expect(dialog.asFragment()).toMatchSnapshot();
    });

    it("calls onFinished with dismissed=true if we dismiss", () => {
        const onFinished = jest.fn();
        const dialog = render(<ConfirmKeyStorageOffDialog onFinished={onFinished} />);

        dialog.getByRole("button", { name: "Yes, dismiss" }).click();

        expect(onFinished).toHaveBeenCalledWith(true);
    });

    it("calls onFinished with dismissed=true if we continue", () => {
        const onFinished = jest.fn();
        const dialog = render(<ConfirmKeyStorageOffDialog onFinished={onFinished} />);

        dialog.getByRole("button", { name: "Go to Settings" }).click();

        expect(onFinished).toHaveBeenCalledWith(false);
    });
});
