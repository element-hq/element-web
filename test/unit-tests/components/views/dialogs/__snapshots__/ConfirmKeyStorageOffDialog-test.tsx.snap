// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmKeyStorageOffDialog renders 1`] = `
<DocumentFragment>
  <div
    class="mx_EncryptionCard"
  >
    <div
      class="mx_EncryptionCard_header"
    >
      <div
        class="_content_o77nw_8 _destructive_o77nw_34"
        data-size="large"
      >
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 17q.424 0 .713-.288A.97.97 0 0 0 13 16a.97.97 0 0 0-.287-.713A.97.97 0 0 0 12 15a.97.97 0 0 0-.713.287A.97.97 0 0 0 11 16q0 .424.287.712.288.288.713.288m0-4q.424 0 .713-.287A.97.97 0 0 0 13 12V8a.97.97 0 0 0-.287-.713A.97.97 0 0 0 12 7a.97.97 0 0 0-.713.287A.97.97 0 0 0 11 8v4q0 .424.287.713.288.287.713.287m0 9a9.7 9.7 0 0 1-3.9-.788 10.1 10.1 0 0 1-3.175-2.137q-1.35-1.35-2.137-3.175A9.7 9.7 0 0 1 2 12q0-2.075.788-3.9a10.1 10.1 0 0 1 2.137-3.175q1.35-1.35 3.175-2.137A9.7 9.7 0 0 1 12 2q2.075 0 3.9.788a10.1 10.1 0 0 1 3.175 2.137q1.35 1.35 2.137 3.175A9.7 9.7 0 0 1 22 12a9.7 9.7 0 0 1-.788 3.9 10.1 10.1 0 0 1-2.137 3.175q-1.35 1.35-3.175 2.137A9.7 9.7 0 0 1 12 22m0-2q3.35 0 5.675-2.325T20 12t-2.325-5.675T12 4 6.325 6.325 4 12t2.325 5.675T12 20"
          />
        </svg>
      </div>
      <h2
        class="_typography_6v6n8_153 _font-heading-sm-semibold_6v6n8_93"
      >
        Are you sure you want to keep key storage turned off?
      </h2>
    </div>
    <span>
      If you sign out of all your devices you will lose your message history and will need to verify all your existing contacts again. 
      <br />
      <a
        href="https://element.io/help#encryption5"
        rel="noreferrer noopener"
        target="_blank"
      >
        Learn more 
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 3h6a1 1 0 1 1 0 2H5v14h14v-6a1 1 0 1 1 2 0v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2"
          />
          <path
            d="M15 3h5a1 1 0 0 1 1 1v5a1 1 0 1 1-2 0V6.414l-6.293 6.293a1 1 0 0 1-1.414-1.414L17.586 5H15a1 1 0 1 1 0-2"
          />
        </svg>
      </a>
    </span>
    <div
      class="mx_EncryptionCard_buttons"
    >
      <button
        class="_button_vczzf_8"
        data-kind="primary"
        data-size="lg"
        role="button"
        tabindex="0"
      >
        Go to Settings
      </button>
      <button
        class="_button_vczzf_8"
        data-kind="secondary"
        data-size="lg"
        role="button"
        tabindex="0"
      >
        Yes, dismiss
      </button>
    </div>
  </div>
</DocumentFragment>
`;
